### PR TITLE
Refactor skill tooling and split out swebench skill

### LINKED ISSUE #298: Adopt a Rust test placement contract and migrate oversized inline test suites
## Problem

The Rust codebase has reached a size where test placement is now hurting implementation readability rather than helping it.

Current repo observations from the local checkout:

- Rust is roughly 103k lines of `code` by `tokei`, or about 117k physical lines.
- Estimated test-related Rust lines are about 55k, or roughly 47% of all Rust physical lines.
- About 90.8% of those test lines live inline inside `src` files rather than under `crates/*/tests/`.
- About 81.5% of Rust files contain tests or are test files.
- The biggest mixed production-plus-test files are concentrated in `alan-runtime` and `alan`.

Notable hotspots:

- `crates/runtime/src/runtime/turn_executor.rs`
- `crates/runtime/src/session.rs`
- `crates/alan/src/daemon/routes.rs`
- `crates/alan/src/daemon/state.rs`
- `crates/runtime/src/tools/sandbox.rs`
- `crates/runtime/src/runtime/virtual_tools.rs`
- `crates/runtime/src/runtime/tool_orchestrator.rs`

The problem is not that inline tests are bad. The problem is that many large async, fixture-heavy, private-access scenario suites are still living at the bottom of already-large implementation files.

## Proposed Contract

A draft spec has been prepared locally at `docs/spec/rust_test_placement_contract.md`.

That contract defines three stable placement tiers for Rust tests:

1. **Inline unit tests**
   - Small, local tests for private helpers and narrow invariants.
2. **Extracted white-box test files**
   - For tests that still need private access but are now large, async, fixture-heavy, or scenario-oriented.
   - Example conventions: `foo_tests.rs`, `foo/tests.rs`, `foo/tests/<topic>.rs`.
3. **Crate-level integration tests**
   - `crates/<crate>/tests/` for black-box API, CLI, route, protocol, smoke, persistence, and live tests.

Key rules from the draft contract:

- Do not widen `pub` or `pub(crate)` solely for test placement.
- Do not keep large async scenario suites inline once they stop being truly local unit tests.
- Prefer extracted white-box test files for large private-access suites.
- Keep black-box behavior tests under `crates/<crate>/tests/`.
- Keep test-only support helpers in test-only locations.

## Why This Matters

We need a repository-wide convention before doing bulk migration work. Without a contract, test moves will be inconsistent and reviewers will keep making one-off placement decisions.

This issue is for adopting the contract and then using it to drive a first migration wave.

## Work Items

- [ ] Land and adopt the Rust test placement contract in `docs/spec/`
- [ ] Link the contract from the spec index and related testing docs where helpful
- [ ] Apply the contract to all new Rust tests going forward
- [ ] Migrate the first wave of oversized mixed files, starting with the hotspot files listed above
- [ ] Move large private-access suites into extracted white-box test files instead of leaving them inline
- [ ] Keep crate-boundary, route, protocol, smoke, and live tests under `crates/<crate>/tests/`
- [ ] Avoid widening production visibility solely to satisfy integration-test placement
- [ ] Add a lightweight review checklist or contributor guidance if that helps enforce the convention

## Suggested First Migration Wave

Start with the largest files where readability gains are obvious and the test blocks are already substantial:

- `crates/runtime/src/runtime/turn_executor.rs`
- `crates/runtime/src/session.rs`
- `crates/alan/src/daemon/routes.rs`
- `crates/alan/src/daemon/state.rs`
- `crates/runtime/src/tools/sandbox.rs`
- `crates/runtime/src/runtime/virtual_tools.rs`
- `crates/runtime/src/runtime/tool_orchestrator.rs`

## Acceptance Signal

This issue is done when new Rust test additions follow the contract by default, and the largest mixed implementation files have begun shrinking through targeted test extraction rather than continuing to grow as production-plus-harness monoliths.

### LINKED ISSUE #301: Prefer Rust CLI/bin surfaces for first-party skill tooling
## Problem Statement

Alan's first-party skill tooling still mixes Rust CLI surfaces with package-local shell and Python helpers. In a few places the script is only wrapping an existing Rust command, while in other places a skill package carries non-trivial Python operator tooling.

That creates a few problems:

- extra interpreter/runtime dependencies for local authoring and eval flows
- duplicated entrypoints around existing Rust surfaces such as `alan` and `alan-skill-tools`
- less typed, less discoverable operator tooling
- friction with Alan's general preference for direct, inspectable commands over opaque script dispatch

We want a clear default for first-party skill tooling: if we do not need shell/Python/TypeScript, do not use them.

## Proposed Solution

Treat this as an authoring/tooling policy and a follow-on implementation track.

Default policy:

- Prefer existing Rust CLI/bin surfaces such as `alan ...` and `alan-skill-tools ...` for first-party skill authoring and evaluation flows.
- Treat package-local scripts as a compatibility/package-private escape hatch rather than the default home for reusable tooling.
- Avoid adding new shell, Python, or TypeScript wrappers when they only forward to existing Rust commands.
- Prefer structured eval manifests (`evals/evals.json`) over new legacy `scripts/eval.py` hooks.
- Keep non-Rust helpers only when the underlying external ecosystem is itself shell- or Python-native, or when a tiny package-private glue step is materially simpler than a new Rust surface.

Initial follow-up scope:

1. Remove or replace thin Python wrappers in `crates/runtime/skills/skill-creator/scripts/` that only forward to `alan` / `alan-skill-tools`.
2. Replace trivial inline Python in shell scripts where a direct shell or existing tool can do the job.
3. Promote the non-trivial SWE-bench preparation helpers under `crates/runtime/skills/repo-coding/scripts/` into Rust CLI/bin tooling.
4. Keep explicit exceptions for external Python-native tooling such as the official SWE-bench harness entrypoint.
5. Review whether legacy `scripts/eval.py` should remain compatibility-only or eventually narrow further for first-party packages.

Documentation has been updated to reflect this policy in:

- `docs/spec/skill_system_contract.md`
- `docs/skill_authoring.md`
- `crates/runtime/skills/skill-creator/SKILL.md`

## Alternatives Considered

- Leave the current mixed model in place and treat language choice as package-local preference.
  This keeps short-term flexibility but preserves wrapper sprawl and inconsistent operator UX.
- Ban scripts entirely.
  This is too rigid because some external ecosystems, test harnesses, and compatibility cases are genuinely script-native.

## Impact and Scope

Likely affected areas:

- `docs/spec/skill_system_contract.md`
- `docs/skill_authoring.md`
- first-party built-in skill packages, especially `skill-creator` and `repo-coding`
- `crates/skill-tools`
- `crates/alan` CLI surfaces for any promoted helper workflows
- tests and prompt-cache assertions that currently reference script resource names

### PR DESCRIPTION
**Pull request by realmorrisliu**
## Summary
- move shared structured eval/review artifact logic into `alan skills` and delete the `crates/skill-tools` crate
- add a standalone built-in `swebench` package with skill-local `bin/`, scripts, eval assets, and a colocated `alan-swebench-tooling` workspace crate
- retarget `repo-coding` and the coding/skill docs so benchmark orchestration lives in `swebench` while `repo-coding` stays the measured coding product
- extract the large `virtual_tools` and `sandbox` white-box test suites into dedicated test modules

Closes #298
Closes #301

## Validation
- `cargo fmt --all`
- `cargo test -p alan-swebench-tooling --no-run`
- `cargo test -p alan-runtime --lib --no-run`
- `cargo test -p alan --test skills_cli_integration_test --no-run`
- `cargo test -p alan-swebench-tooling swebench_lite::tests:: --lib`
- `cargo test -p alan-runtime skills::capability_view::tests::builtin_swebench_package_exposes_operator_resources --lib`
- `cargo test -p alan --test skills_cli_integration_test skills_aggregate_benchmark_and_generate_review_rebuild_artifacts`
- `bash -n crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces crates/runtime/skills/swebench/bin/swebench-lite-materialize-subset crates/runtime/skills/swebench/scripts/check_swebench_harness_env.sh crates/runtime/skills/swebench/scripts/run_swebench_full_steward_case.sh crates/runtime/skills/swebench/scripts/run_swebench_full_steward_subset.sh crates/runtime/skills/swebench/scripts/score_swebench_predictions.sh crates/runtime/skills/swebench/scripts/setup_swebench_harness_env.sh crates/runtime/skills/skill-creator/bin/aggregate-benchmark crates/runtime/skills/skill-creator/bin/generate-review crates/runtime/skills/skill-creator/scripts/aggregate_benchmark.py crates/runtime/skills/skill-creator/scripts/generate_review.py`
- `git diff --check`

### CODE REVIEW

File: crates/runtime/skills/swebench/tooling/src/swebench_lite.rs [hunk 1]
```diff
@@ -0,0 +1,1268 @@
...
+    workspace_dir: &Path,
+    repo: &str,
+    base_commit: &str,
+    github_root: &str,
+    mirror_path: &Path,
+    reuse_existing_workspaces: bool,
+) -> Result<WorkspacePreparationOutcome> {
+    let mut recreated = false;
+    if workspace_dir.exists() {
+        if reuse_existing_workspaces && existing_workspace_matches(workspace_dir, base_commit)? {
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/swebench/tooling/src/swebench_lite.rs]: Fixed in 163bc0e. `prepare_workspace_instance` now treats `existing_workspace_matches(...)` failures as a signal to recreate the owned workspace instead of failing the instance, so interrupted or non-git leftovers self-heal again under `--reuse-existing-workspaces`. I also added a regression test covering a stale non-git workspace directory.

File: crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces:1]: Fixed in 163bc0e. I set the executable bit on both new `swebench/bin/*` entrypoints, and I also fixed the same issue for the new `skill-creator/bin/*` wrappers since they are documented as direct package-local command surfaces too.

File: crates/runtime/skills/skill-creator/bin/aggregate-benchmark [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+script_dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+repo_root="$(cd "$script_dir/../../../../.." && pwd)"
+
+if command -v alan >/dev/null 2>&1; then
+    exec alan skills aggregate-benchmark "$@"
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/skill-creator/bin/aggregate-benchmark:22]: Fixed in 0b444f1. Both `skill-creator/bin/*` wrappers now prefer the repo-local `target/{debug,release}/alan` build first, then fall back to `cargo run`, so a stale global `alan` on PATH can no longer intercept these package-local entrypoints.

File: crates/runtime/skills/skill-creator/bin/generate-review [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/skill-creator/bin/generate-review:1]: Fixed in 163bc0e. I set the executable bit on both `skill-creator/bin/*` wrappers together with the new `swebench/bin/*` entrypoints so the documented package-local command surfaces are directly runnable.

File: crates/runtime/skills/skill-creator/bin/aggregate-benchmark [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    "$repo_root/target/debug/alan" \
+    "$repo_root/target/release/alan"
+do
+    if [[ -x "$candidate" ]]; then
+        exec "$candidate" skills aggregate-benchmark "$@"
+    fi
+done
+
+cd "$repo_root"
+exec cargo run -p alan -- skills aggregate-benchmark "$@"
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/skill-creator/bin/aggregate-benchmark]: Fixed in ed839e2. The skill-creator wrappers now keep the repo-local target/cargo path for source checkouts, but only run cargo when the derived root is actually an Alan source checkout; materialized package invocations fall back to `alan` on PATH. Added `package_bin_wrappers_fall_back_to_path_outside_source_tree` coverage for this packaged case, including the sibling `generate-review` wrapper.

File: crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    "$repo_root/target/debug/swebench-lite-prepare-workspaces" \
+    "$repo_root/target/release/swebench-lite-prepare-workspaces"
+do
+    if [[ -x "$candidate" ]]; then
+        exec "$candidate" "$@"
+    fi
+done
+
+cd "$repo_root"
+exec cargo run -p alan-swebench-tooling --bin swebench-lite-prepare-workspaces -- "$@"
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces]: Fixed in ed839e2. Both swebench bin wrappers now keep repo-local target/cargo execution in source checkouts and fall back to the matching PATH binary when invoked from a materialized package outside the source tree. The same-name PATH fallback also guards against resolving back to the wrapper itself. Added regression coverage for both prepare and materialize wrappers.

File: crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    "$repo_root/target/release/$tool_name"
+do
+    if [[ -x "$candidate" ]]; then
+        exec "$candidate" "$@"
+    fi
+done
+
+if [[ -f "$repo_root/Cargo.toml" && -f "$repo_root/crates/runtime/skills/swebench/tooling/Cargo.toml" ]] && command -v cargo >/dev/null 2>&1; then
+    cd "$repo_root"
+    exec cargo run -p alan-swebench-tooling --bin "$tool_name" -- "$@"
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/swebench/bin/swebench-lite-prepare-workspaces:40]: Fixed in 833c7ca. The swebench package wrappers now fall back to `alan skills <entrypoint>` when invoked from a materialized package with no source checkout or helper binary, so normal installs that only expose `alan` can run the package-local bin surface. I also kept and hardened the same-name helper fallback by scanning PATH and skipping the wrapper itself.

File: crates/runtime/skills/swebench/tooling/src/swebench_lite.rs [hunk 8]
```diff
@@ -0,0 +1,1321 @@
...
+    if let Ok(canonical) = fs::canonicalize(path) {
+        return Ok(canonical);
+    }
+    let absolute = normalize_path(&if path.is_absolute() {
+        path.to_path_buf()
+    } else {
+        std::env::current_dir()
+            .context("Failed to resolve current directory")?
+            .join(path)
+    });
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/swebench/tooling/src/swebench_lite.rs:1043]: Fixed in 833c7ca. `resolve_path` now expands `~` and `~/...` from HOME/USERPROFILE before canonicalization and normalization, so workspace-map paths no longer resolve under the current working directory. Added regression coverage for a non-existing home-relative path.

File: crates/runtime/skills/swebench/tooling/src/swebench_lite.rs [hunk 9]
```diff
@@ -0,0 +1,1363 @@
...
+    run_git(
+        vec![
+            OsString::from("clone"),
+            OsString::from("--mirror"),
+            mirror_path_arg(&clone_url),
+            mirror_path.as_os_str().to_owned(),
+        ],
+        None,
+    )?;
+    Ok(true)
```
[the remark this answers is not in the document]
Comment (realmorrisliu) [reply crates/runtime/skills/swebench/tooling/src/swebench_lite.rs]: Fixed in 2cc1d8c. `ensure_repo_mirror` now returns `true` only when an existing invalid mirror was removed and cloned again, and returns `false` for first-time mirror creation. I also extended the local git repository preparation test to cover both report semantics: first-run clones leave `recreated_mirrors` empty, while a stale non-bare mirror is reported as recreated on rerun.
